Sem. 03/Solutions/Task01.cpp

```diff
@@ -0,0 +1,183 @@
+#include <iostream>
+#include <fstream>
+
+const char* UNSORTED_FILE_NAME = "unsortedDeliveries.bin";
+const char* SORTED_FILE_NAME = "sortedDeliveries.bin";
+const char* ERROR_MESSAGE = "Error";
+
+const int ADDRESS_MAX_LENGTH = 25;
+
+struct Delivery {
+    double price;
+    char address[ADDRESS_MAX_LENGTH];
+};
+
+int stringLength(const char* string) {
+    int length = 0;
+
+    for (int i = 0; string[i] != '\0'; ++i) {
+        length++;
+    }
+
+    return length;
+}
+
+int minStringLength(const char* string1, const char* string2) {
+    int minLength = 0;
+
+    for (int i = 0; string1[i] != '\0' || string2[i] != '\0'; ++i) {
+        minLength++;
+    }
+
+    return minLength;
+}
+
+bool isSecondStringLarger(const char* string1, const char* string2) {
+    int minLength = minStringLength(string1, string2);
+
+    for (int i = 0; i < minLength; ++i) {
+        if (string1[i] == string2[i]) {
+            continue;
+        }
+
+        return string1[i] < string2[i];
+    }
+
+    return false;
+}
+
+size_t getFileSize(std::ifstream& in) {
+
+    size_t currentPosition = in.tellg();
+    in.seekg(0, std::ios::end);
+    size_t fileSize = in.tellg();
+    in.seekg(currentPosition, std::ios::beg);
+
+    in.clear();
+
+    return fileSize;
+}
+
+namespace deliveryFunctions {
+
+    void assignAddress(Delivery& delivery, const char* address) {
+        int length = stringLength(address) + 1;
+
+        for (int i = 0; i < length; ++i) {
+            delivery.address[i] = address[i];
+        }
+    }
+
+    void swapDeliveries(Delivery& delivery1, Delivery& delivery2) {
+        Delivery temp = delivery1;
+        delivery1 = delivery2;
+        delivery2 = temp;
+    }
+
+    void selectionSort(Delivery* deliveryArray, const size_t arraySize) {
+        for (int i = 0; i < arraySize - 1; ++i) {
+            int minIndex = i;
+
+            for (int j = i + 1; j < arraySize; ++j) {
+
+                if (isSecondStringLarger(deliveryArray[j].address, deliveryArray[minIndex].address)) {
+                    minIndex = j;
+                }
+
+            }
+
+            if (minIndex != i) {
+                swapDeliveries(deliveryArray[i], deliveryArray[minIndex]);
+            }
+        }
+    }
+
+    void initializeDelivery(Delivery& delivery, const double price, const char* address) {
```

Comment:
no need for const for primitive types (double).

Reply:
Isn't it good courtesy to explicitly state a parameter is constant if you're not changing it, whatever the type? That's what I would assume.

Reply:
The functions are short enough to see that the parameter is not changed.
Since the parameter is passed as a copy, nobody cares if it is modified in this function.